File: 026주차/게임맵최단거리/woosteelz.py
# 프로그래머스 게임 맵 최단거리 (Level 2)
from collections import deque


def solution(maps):
    dir_x = [1, -1, 0, 0]
    dir_y = [0, 0, 1, -1]
    N = len(maps)
    M = len(maps[0])
    visited = [[0 for _ in range(M)] for _ in range(N)]

    queue = deque()
    queue.append([0, 0])
    visited[0][0] = 1
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            temp_x = x + dir_x[i]
            temp_y = y + dir_y[i]
            if 0 <= temp_x < N and 0 <= temp_y < M and not visited[temp_x][temp_y] and maps[temp_x][temp_y] == 1:
                queue.append([temp_x, temp_y])
                visited[temp_x][temp_y] = 1 + visited[x][y]
    if not visited[N-1][M-1]:
        return -1

    return visited[N-1][M-1]


maps = [[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [
        1, 0, 1, 1, 1], [1, 1, 1, 0, 0], [0, 0, 0, 0, 1]]
print(solution(maps))

File: 026주차/게임맵최단거리/woosteelz.js
const solution = (maps) => {
    let ans = 0;
    const dirX = [0, -1, 1, 0]
    const dirY = [1, 0, 0, -1]
    const n = maps.length
    const m = maps[0].length
    const visited = [...maps].map((r) => r.map((c) => 0));

    const queue = []
    queue.push([0, 0])
    visited[0][0] = 1

    while (queue.length) {
        const [currX, currY] = queue.shift()
        for (let i = 0; i < 4; i++) {
            const [nextX, nextY] = [currX + dirX[i], currY + dirY[i]]
            if (0 <= nextX && nextX < n && 0 <= nextY && nextY < m && visited[nextX][nextY] === 0 && maps[nextX][nextY] === 1) {
                queue.push([nextX, nextY])
                visited[nextX][nextY] = visited[currX][currY] + 1
            }
        }
    }

    return !visited[n - 1][m - 1] ? -1 : visited[n - 1][m - 1]
}

// console.log는 전 부 지 우 자 

console.log(solution([[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 1], [0, 0, 0, 0, 1]]))